src/Debug.c

```diff
@@ -24,31 +30,92 @@ void Ros_Debug_Init()
+    int count = 0;
+    int* socket = ros_debugPorts.debugSocket;
+    struct sockaddr_in* sin = ros_debugPorts.destAddr;
+
+    bzero(&ros_debugPorts, sizeof(ros_debugPorts));
+
+    for (int i = 1; i <= MAX_NETWORK_PORTS; i++)
+    {
+
+        status = Ros_mpNICData(i, &ip_be, &subnetmask_be, mac, &gateway_be);
+
+        if (status == OK)
+        {
+            socket[count] = mpSocket(AF_INET, SOCK_DGRAM, 0);
+            Ros_setsockopt(socket[count], SOL_SOCKET, SO_BROADCAST, (char*)&broadcastVal, sizeof(broadcastVal));
+            sin[count].sin_addr.s_addr = ip_be | (~subnetmask_be);
+            sin[count].sin_family = AF_INET;
+            sin[count].sin_port = mpHtons(DEBUG_UDP_PORT_NUMBER);
+            count++;
+        }
+    }
+    ros_debugPorts.enabledPortCount = count;
+}
+
+void Ros_Debug_SetFromConfig()
+{
+    ULONG ip_be;
+    ULONG subnetmask_be;
+    ULONG gateway_be;
+    int broadcastVal = 1;
+    UCHAR mac[6];
+    STATUS status;
+    char message[ERROR_MSG_MAX_SIZE];
+    int* socket = ros_debugPorts.debugSocket;
+    struct sockaddr_in* sin = ros_debugPorts.destAddr;
 
-    ros_DebugSocket = mpSocket(AF_INET, SOCK_DGRAM, 0);
-    Ros_setsockopt(ros_DebugSocket, SOL_SOCKET, SO_BROADCAST, (char*)&broadcastVal, sizeof(broadcastVal));
+    bzero(&ros_debugPorts, sizeof(ros_debugPorts));
 
-    Ros_mpNICData(ROS_USER_LAN1, &ip_be, &subnetmask_be, mac, &gateway_be);
+    status = Ros_mpNICData(g_nodeConfigSettings.debug_broadcast_port, &ip_be, &subnetmask_be, mac, &gateway_be);
 
-    ros_debug_destAddr1.sin_addr.s_addr = ip_be | (~subnetmask_be);
-    ros_debug_destAddr1.sin_family = AF_INET;
-    ros_debug_destAddr1.sin_port = mpHtons(DEBUG_UDP_PORT_NUMBER);
+    if (status == OK)
+    {
+        socket[0] = mpSocket(AF_INET, SOCK_DGRAM, 0);
+        Ros_setsockopt(socket[0], SOL_SOCKET, SO_BROADCAST, (char*)&broadcastVal, sizeof(broadcastVal));
+        sin[0].sin_addr.s_addr = ip_be | (~subnetmask_be);
+        sin[0].sin_family = AF_INET;
+        sin[0].sin_port = mpHtons(DEBUG_UDP_PORT_NUMBER);
+        ros_debugPorts.enabledPortCount = 1;
+    }
+    else
+    {
+        int ret = snprintf(message, ERROR_MSG_MAX_SIZE, "Enable LAN port %d for debug", g_nodeConfigSettings.debug_broadcast_port);
+        if (0 < ret && ret <= ERROR_MSG_MAX_SIZE)
+        {
+            mpSetAlarm(ALARM_ASSERTION_FAIL, message, SUBCODE_DEBUG_INIT_FAIL_MP_NICDATA);
+        }
+        else
+        {
+            mpSetAlarm(ALARM_ASSERTION_FAIL, "Enable debug LAN port from cfg", SUBCODE_DEBUG_INIT_FAIL_MP_NICDATA);
+        }
+        g_nodeConfigSettings.debug_broadcast_enabled = FALSE;
+    }
 }
 
 void Ros_Debug_BroadcastMsg(char* fmt, ...)
 {
     char str[MAX_DEBUG_MESSAGE_SIZE];
     va_list va;
 
+    if (g_nodeConfigSettings.debug_broadcast_enabled && ros_debugPorts.enabledPortCount == 0)
+    {
+        Ros_Debug_Init();
+    }
+
+    if (!g_nodeConfigSettings.debug_broadcast_enabled && !g_nodeConfigSettings.log_to_stdout)
```

Comment:
this might also be a nice place to explain why this will allow control flow to continue here if `g_nodeConfigSettings.log_to_stdout == TRUE`?

(because we can reuse all the machinery after this to format the message and as `ros_debugPorts.enabledPortCount` is most likely `0`, nothing will be sent out over any of the LAN ports, but it will be printed to the controller's console. IIUC).

Reply:
@jimmy-mcelwain: I'm wondering whether this works as intended (specifically this line).

Could it be this prevents logs from getting broadcast until at least the config file has been parsed? Both `debug_broadcast_enabled` and `log_to_stdout` will be `FALSE` in that case I believe.

Looking at the debug log output of my YRC1, it seems to be missing at least these printouts:

https://github.com/Yaskawa-Global/motoros2/blob/c61fb86e2cf31ff894189efd8a36a2f1fc5d1161/src/main.c#L60-L64

Reply:
I think that you're right. Maybe if we call `Ros_ConfigFile_SetAllDefaultValues()` to set `debug_broadcast_enabled` to `TRUE` in `main` before any calls to `Ros_Debug_BroadcastMsg()`, then we would be able to broadcast those messages. I would need to think it through more though, that might require more changes.